### ISSUE TITLE
upgrade commons-lang to 3.5

### ISSUE THREAD

**Issue by venusdrogon** [venusdrogon/feilong-platform#24]
http://commons.apache.org/proper/commons-lang/release-notes/RELEASE-NOTES-3.5.txt

  Apache Commons Lang
                               Version 3.5
                              Release Notes

**INTRODUCTION**:

 Commons Lang 3.5 at least requires **Java 6.0**.

For the advice on upgrading from 2.x to 3.x, see the following page:

```
http://commons.apache.org/lang/article3_0.html
```

**Comment by venusdrogon** [comment]
# HIGHLIGHTS

Some of the highlights in this release include:

o Added Java 9 detection to org.apache.commons.lang3.**SystemUtils**.
o Support for shifting and swapping elements in org.apache.commons.lang3.ArrayUtils.
o New methods for generating random strings from different character classes including alphabetic, alpha-numeric and ASCII added to  org.apache.commons.lang3.**RandomStringUtils**.
o Numerous extensions to org.apache.commons.lang3.StringUtils including null safe compare variants, more remove and replace variants, rotation and truncation.
o Added org.apache.commons.lang3.**ThreadUtils** - a utility class to work with instances of java.lang.Thread and java.lang.ThreadGroup.
o Added annotations **@EqualsExclude**, **@HashCodeExclude** and **@ToStringEclude** to mark fields which should be ignored by the reflective builders in the    org.apache.commons.lang3.builder package.
o Support for various modify and retrieve value use cases added to the classes  in org.apache.commons.lang3.mutable.

**Comment by venusdrogon** [comment]
# JAVA 9 SUPPORT

Java 9 introduces a new version-string scheme. Details of this new scheme are documented in JEP-223 (http://openjdk.java.net/jeps/223). 

In order to support JEP-223 two classes had to be changed:

o org.apache.commons.lang3.**JavaVersion**
    deprecated enum constant JAVA_1_9
    introduced enum constant JAVA_9

o org.apache.commons.lang3.*\* #SystemUtils**
    deprecated constant IS_JAVA_1_9
    introduced constant IS_JAVA_9

For more information see LANG-1197
(https://issues.apache.org/jira/browse/LANG-1197). All other APIs are expected to work with Java 9.

# BUILDING ON JAVA 9

Java 8 introduced the Unicode Consortium's Common Locale Data Repository as alternative source for locale data. Java 9 will use the CLDR provider as
default provider for locale data (see http://openjdk.java.net/jeps/252). This causes an number of locale-sensitive test in Commons Lang to fail. In order
to build Commons Lang 3.5 on Java 9, the locale provider has to be set to 'JRE':

```
mvn -Djava.locale.providers=JRE clean install
```

We are currently investigating ways to support building on Java 9 without further configuration. 

For more information see: https://issues.apache.org/jira/browse/LANG-1265

**Comment by venusdrogon** [comment]
# NEW FEATURES
- LANG-1275: Added a tryAcquire() method to TimedSemaphore.
- LANG-1255: Add DateUtils.toCalendar(Date, TimeZone). 
- LANG-1023: Add WordUtils.wrap overload with customizable breakable character. 
- LANG-787:  Add method removeIgnoreCase(String, String) to StringUtils. 
- LANG-1224: Extend RandomStringUtils with methods that generate strings between a min and max length.
- LANG-1257: Add APIs StringUtils.wrapIfMissing(String, char|String). 
- LANG-1253: Add RandomUtils#nextBoolean() method. 
- LANG-1085: Add a circuit breaker implementation. 
- LANG-1013: Add StringUtils.truncate().
- LANG-1195: Enhance MethodUtils to allow invocation of private methods. 
- LANG-1189: Add getAndIncrement/getAndDecrement/getAndAdd/incrementAndGet/decrementAndGet/addAndGet in Mutable\* classes.
- LANG-1225: Add RandomStringUtils#randomGraph and #randomPrint which match             corresponding regular expression class. 
- LANG-1223: Add StopWatch#getTime(TimeUnit). 
- LANG-781:  Add methods to ObjectUtils class to check for null elements in the  array. 
- LANG-1228: Prefer Throwable.getCause() in ExceptionUtils.getCause(). 
- LANG-1233: DiffBuilder add method to allow appending from a DiffResult.  
- LANG-1168: Add SystemUtils.IS_OS_WINDOWS_10 property.
- LANG-1115: Add support for varargs in ConstructorUtils, MemberUtils, and MethodUtils. 
- LANG-1134: Add methods to check numbers against NaN and inifinite to  Validate. 
- LANG-1220: Add tests for missed branches in DateUtils. 
- LANG-1146: z/OS identification in SystemUtils.  
- LANG-1192: FastDateFormat support of the week-year component (uppercase 'Y'). 
- LANG-1169: Add StringUtils methods to compare a string to multiple strings. 
- LANG-1185: Add remove by regular expression methods in StringUtils.
- LANG-1139: Add replace by regular expression methods in StringUtils.
- LANG-1171: Add compare methods in StringUtils.
- LANG-1174: Add sugar to RandomUtils. 
- LANG-1154: FastDateFormat APIs that use a StringBuilder. 
- LANG-1149: Ability to throw checked exceptions without declaring them. 
- LANG-1153: Implement ParsePosition api for FastDateParser.
- LANG-1137: Add check for duplicate event listener in EventListenerSupport. 
- LANG-1135: Add method containsAllWords to WordUtils.
- LANG-1132: ReflectionToStringBuilder doesn't throw IllegalArgumentException when the constructor's object param is null. 
- LANG-701:  StringUtils join with var args. 
- LANG-1105: Add ThreadUtils - A utility class which provides helper methods   related to java.lang.Thread  Issue: LANG-1105. 
- LANG-1031: Add annotations to exclude fields from ReflectionEqualsBuilder, ReflectionToStringBuilder and ReflectionHashCodeBuilder. 
- LANG-1127: Use JUnit rules to set and reset the default Locale and TimeZone.
- LANG-1119: Add rotate(string, int) method to StringUtils. 
- LANG-1099: Add swap and shift operations for arrays to ArrayUtils. 
- LANG-1050: Change nullToEmpty methods to generics. 
- LANG-1074: Add a method to ArrayUtils for removing all occurrences of a given element  Issue: LANG-1074.

**Comment by venusdrogon** [comment]
# FIXED BUGS
- LANG-1261: ArrayUtils.contains returns false for instances of subtypes.
- LANG-1252: Rename NumberUtils.isNumber, isCreatable to better reflect createNumber. Also, accommodated for "+" symbol as prefix in isCreatable and isNumber. 
- LANG-1230: Remove unnecessary synchronization from registry lookup in EqualsBuilder and HashCodeBuilder.
- LANG-1214: Handle "void" in ClassUtils.getClass(). 
- LANG-1250: SerializationUtils#deserialize has unnecessary code and a comment for that. 
- LANG-1190: TypeUtils.isAssignable throws NullPointerException when fromType has type variables and toType generic superclass specifies type variable. 
- LANG-1226: StringUtils#normalizeSpace does not trim the string anymore.  
- LANG-1251: SerializationUtils.ClassLoaderAwareObjectInputStream should use  static initializer to initialize primitiveTypes map.
- LANG-1248: FastDatePrinter Memory allocation regression. 
- LANG-1018: Fix precision loss on NumberUtils.createNumber(String). 
- LANG-1199: Fix implementation of StringUtils.getJaroWinklerDistance(). 
- LANG-1244: Fix dead links in StringUtils.getLevenshteinDistance() javadoc.
- LANG-1242: "\u2284":"?" mapping missing from EntityArrays#HTML40_EXTENDED_ESCAPE.
- LANG-901:  StringUtils#startsWithAny/endsWithAny is case sensitive - documented as case insensitive. 
- LANG-1232: DiffBuilder: Add null check on fieldName when appending Object or Object[]. 
- LANG-1178: ArrayUtils.removeAll(Object array, int... indices) should do the clone, not its callers. 
- LANG-1120: StringUtils.stripAccents should remove accents from "Ł" and "ł". 
- LANG-1205: NumberUtils.createNumber() behaves inconsistently with  NumberUtils.isNumber(). 
- LANG-1222: Fix for incorrect comment on StringUtils.containsIgnoreCase method. 
- LANG-1221: Fix typo on appendIfMissing javadoc. 
- LANG-1202: parseDateStrictly does't pass specified locale. 
- LANG-1219: FastDateFormat doesn't respect summer daylight in some localized  strings.
- LANG-1175: Remove Ant-based build.
- LANG-1194: Limit max heap memory for consistent Travis CI build.
- LANG-1186: Fix NullPointerException in FastDateParser$TimeZoneStrategy.
- LANG-1193: ordinalIndexOf("abc", "ab", 1) gives incorrect answer of -1   (correct answer should be 0); revert fix for LANG-1077. 
- LANG-1002: Several predefined ISO FastDateFormats in DateFormatUtils are  incorrect.
- LANG-1152: StringIndexOutOfBoundsException or field over-write for large year  fields in FastDateParser. 
- LANG-1141: StrLookup.systemPropertiesLookup() no longer reacts on changes on system properties.
- LANG-1147: EnumUtils *BitVector issue with more than 32 values Enum.
- LANG-1059: Capitalize javadoc is incorrect. 
- LANG-1122: Inconsistent behavior of swap for malformed inputs. 
- LANG-1130: Fix critical issues reported by SonarQube.
- LANG-1131: StrBuilder.equals(StrBuilder) doesn't check for null inputs.
- LANG-1128: JsonToStringStyle doesn't handle chars and objects correctly. 
- LANG-1126: DateFormatUtilsTest.testSMTP depends on the default Locale.
- LANG-1123: Unit test FastDatePrinterTimeZonesTest needs a timezone set.  
- LANG-916:  DateFormatUtils.format does not correctly change Calendar TimeZone in certain situations. 
- LANG-1116: DateUtilsTest.testLang530 fails for some timezones. 
- LANG-1114: TypeUtils.ParameterizedType#equals doesn't work with wildcard  types. 
- LANG-1118: StringUtils.repeat('z', -1) throws NegativeArraySizeException. 
- LANG-1111: Fix FindBugs warnings in DurationFormatUtils.
- LANG-1162: StringUtils#equals fails with Index OOBE on non-Strings with  identical leading prefix..
- LANG-1163: There are no tests for CharSequenceUtils.regionMatches.
- LANG-1200: Fix JavaDoc of StringUtils.ordinalIndexOf. 
- LANG-1191: Incorrect Javadoc  StringUtils.containsAny(CharSequence, CharSequence...).

**Comment by venusdrogon** [comment]
# CHANGES
- LANG-1197: Prepare Java 9 detection.
- LANG-1262: CompareToBuilder.append(Object, Object, Comparator) method is too big to be inlined.
- LANG-1259: JavaDoc for ArrayUtils.isNotEmpty() is slightly misleading. 
- LANG-1247: FastDatePrinter generates extra Date objects. 
- LANG-1229: HashCodeBuilder.append(Object,Object) is too big to be inlined, which prevents whole builder to be scalarized. 
- LANG-1243: Simplify ArrayUtils removeElements by using new decrementAndGet()  method.
- LANG-1240: Optimize BitField constructor implementation. 
- LANG-1206: Improve CharSetUtils.squeeze() performance. 
- LANG-1176: Improve ArrayUtils removeElements time complexity to O(n). 
- LANG-1234: getLevenshteinDistance with a threshold: optimize implementation             if the strings lengths differ more than the threshold.
- LANG-1151: Performance improvements for NumberUtils.isParsable.
- LANG-1218: EqualsBuilder.append(Object,Object) is too big to be inlined,which prevents whole builder to be scalarized. 
- LANG-1210: StringUtils#startsWithAny has error in Javadoc.
- LANG-1208: StrSubstitutor can preserve escapes.
- LANG-1182: Clarify JavaDoc of StringUtils.containsAny().
- LANG-1183: Making replacePattern/removePattern methods null safe in StringUtils.
- LANG-1057: Replace StringBuilder with String concatenation for better  optimization. 
- LANG-1075: Deprecate `SystemUtils.FILE_SEPARATOR` and `SystemUtils.PATH_SEPARATOR`.
- LANG-979:  TypeUtils.parameterizeWithOwner - wrong format descriptor for invalid number of type parameters".
- LANG-1112: MultilineRecursiveToStringStyle largely unusable due to being package-private.
- LANG-1058: StringUtils.uncapitalize performance improvement. 
- LANG-1069: CharSet.getInstance documentation does not clearly explain how  to include negation character in set.
- LANG-1107: Fix parsing edge cases in FastDateParser.
- LANG-1273: Added new property IS_OS_MAC_OSX_EL_CAPITAN in SystemUtils.

**Comment by venusdrogon** [comment]
# COMPATIBILITY

Apache Commons Lang 3.5 is binary compatible with the 3.4 release. Users should not experience any problems when upgrading vom 3.4 to 3.5.

There has been an addition to the org.apache.commons.lang3.time.**DatePrinter** interface:

o Added method 'public boolean **parse**(java.lang.String, java.text.ParsePosition,java.util.Calendar)'
o Added method 'public java.lang.Appendable **format**(long, java.lang.Appendable)'
o Added method 'public java.lang.Appendable **format**(java.util.Date,java.lang.Appendable)'
o Added method 'public java.lang.Appendable **format**(java.util.Calendar,java.lang.Appendable)'

For this reason 3.5 is not strictly source compatible to 3.4. Since the DatePrinter interface is not meant to be implemented by clients, this change it not considered to cause any problems.

**Comment by venusdrogon** [comment]
fix by c3922585ab3f4fbaab196388a10de6ff7b27e880
